s3/src/main/java/com/yahoo/ycsb/db/S3Client.java

```diff
@@ -0,0 +1,348 @@
+import com.amazonaws.auth.*;
+import com.amazonaws.services.s3.model.ObjectMetadata;
+import com.amazonaws.services.s3.model.PutObjectResult;
+import com.amazonaws.services.s3.model.S3Object;
+import com.amazonaws.services.s3.model.GetObjectRequest;
+import com.amazonaws.ClientConfiguration;
+import com.amazonaws.regions.Region;
+import com.amazonaws.regions.Regions;
+import com.amazonaws.services.s3.model.DeleteObjectRequest;
+import com.amazonaws.services.s3.model.ObjectListing;
+import com.amazonaws.services.s3.model.S3ObjectSummary;
+
+/**
+ * S3 Storage client for YCSB framework.
+ * 
+ * Properties to set:
+ * 
+ * s3.accessKeyId=access key S3 aws
+ * s3.secretKey=secret key S3 aws
+ * s3.endPoint=s3.amazonaws.com
+ * s3.region=us-east-1
+ * The parameter table is the name of the Bucket where to upload the files. 
+ * This must be created before to start the benchmark
+ * The size of the file to upload is determined by two parameters:
+ * - fieldcount this is the number of fields of a record in YCSB
+ * - fieldlength this is the size in bytes of a single field in the record
+ * together these two parameters define the size of the file to upload, 
+ * the size in bytes is given by the fieldlength multiplied by the fieldcount.
+ * The name of the file is determined by the parameter key. 
+ *This key is automatically generated by YCSB.
+ *
+ * @author ivanB1975
+ */
+public class S3Client extends DB {
+ 
+  private static String accessKeyId;
+  private static String secretKey;
+  private static String endPoint;
+  private static String region;
+  private static String maxErrorRetry;
+  private static BasicAWSCredentials s3Credentials;
+  private static AmazonS3Client s3Client;
+  private static ClientConfiguration clientConfig;
+  private static final AtomicInteger INIT_COUNT = new AtomicInteger(0);
+
+  /**
+  * Cleanup any state for this storage.
+  * Called once per S3 instance; 
+  */
+  @Override
+  public void cleanup() throws DBException {
+    if (INIT_COUNT.decrementAndGet() == 0) {
+      try {
+        this.s3Client.shutdown();
+      } catch (Exception e){
+        e.printStackTrace();
+      } finally {
+        this.s3Client = null;
+      }
+    }
+  }
+  /**
+  * Delete a file from S3 Storage.
+  * 
+  * @param bucket
+  *            The name of the bucket
+  * @param key
+  * The record key of the file to delete.
+  * @return Zero on success, a non-zero error code on error. See the
+  * {@link DB} class's description for a discussion of error codes.
+  */
+  @Override
+  public int delete(String bucket, String key) {
+    try {
+      this.s3Client.deleteObject(new DeleteObjectRequest(bucket, key));
+    } catch (Exception e){
+      e.printStackTrace();
+      return 1;
+    }
+    return 0;
+  }
+  /**
+  * Initialize any state for the storage.
+  * Called once per S3 instance; If the client is not null it is re-used.
+  */
+  @Override
+  public void init() throws DBException {
+    INIT_COUNT.incrementAndGet();
```

Comment:
I changed the code taking as example the asnc mongodb binding. I look at the number of threads created and I throw a database exception if the number of threads is bigger than the recordcount or the operation count. The banchmark is ended correctly but the exceeding threads fails with the DBException error. It seems to work correctly.